[docs] add repository plugin (#108)
moodle/devdocs: docs/apis/plugintypes/repository/index.mdx

---
title: Repository plugins
tags:
  - Repositories
  - Plugins
---

import {
    DbAccessPHP,
    Lang,
    Lib,
    VersionPHP,
} from '../../_files';

import {
    CodeBlock,
    Tabs,
    TabItem
} from '@site/src/components';

Repository plugin allow Moodle to bring contents into Moodle from external repositories.

### Prerequisites

Before starting coding, it is necessary to know how to use repository administration pages and how to use the file picker.

> The 3 different parts to write in order to implement a full repository:

Reviewer: Only two parts now listed.

Author: Fixed

2. File picker integration - The core of your plugin, it will manage communication between Moodle and the repository service, and also the file picker display.

## File structure

Repository plugins are located in the `/repository` directory.

Each plugin is in a separate subdirectory and consists of a number of _mandatory files_ and any other files the developer is going to use.

<details>
  <summary>View an example directory layout for the `repository_pluginname` plugin.</summary>

```console
 repository/pluginname/
 |-- db
 |   `-- access.php
 |-- lang
 |   `-- en
 |       `-- repository_pluginname.php
 |-- lib.php
 |-- pix
 |   `-- icon.png
 `-- version.php
```

</details>

Some of the important files for the repository plugintype are described below. See the [common plugin files](../commonfiles) documentation for details of other files which may be useful in your plugin.

### version.php

<VersionPHP
    plugintype="repository"
/>

### lang/en/repository_pluginname.php

export const langExample = `$string['pluginname']= 'Example repository';
$string['configplugin'] = 'Configuration for Example repository';

> $string['pluginname_help'] = 'A repository description';`;

Author: The mdx lint is complaining

> $string['pluginname_help'] = 'A repository description';`;

Reviewer: Should be fine to try again?

<Lang
    plugintype="repository"
    pluginname="pluginname"
    example={langExample}
/>

### lib.php

This file contains the main repository class definition. This class must extend the `repository` class. Once this class is created, the plugin will be able to override the parent class methods to alter the way Moodle interact with the repository. See [Administration API](#administration-apis) for more information.

<details>
  <summary>View example</summary>
  <div>

import RepoLib from '!!raw-loader!./_examples/lib.php';

<CodeBlock language="js" title="repository/pluginname/lib.php">{RepoLib}</CodeBlock>

  </div>

> </details>

Author: done

### db/access.php

<!-- cspell:disable -->

export const accessExample = `$capabilities = [
    // Ability to use the plugin.
    'repository/pluginname:view' => [
        'captype' => 'read',
        'contextlevel' => CONTEXT_MODULE,
        'archetypes' => [
            'coursecreator' => CAP_ALLOW,
            'editingteacher' => CAP_ALLOW,
            'manager' => CAP_ALLOW
        ]
    ],
];`;

> <!-- cspell:enable -->

Author: The mdx lint is complaining

> <!-- cspell:enable -->

Author: I think it is a good idea. Now access is a separate file

<DbAccessPHP
    plugintype="repository"
    pluginname="pluginname"
    example={accessExample}
/>

## Install a new repository plugin

1. Copy the plugin code of your plugin in a **/repository/** subfolder. For exmaple, **/repository/pluginname**
2. Login as admin on your website and run upgrade
3. Open Site Administration->Plugins->Repositories->Manage Repositories and make your repository **Enabled and visible**

> 

Author: done

## Administration APIs

### Fixed settings

These are settings that are hard-coded into your repository plugin and can only be updated by changing the plugin code.

> #### supported_returntypes(): int

Author: Done

Return any combination of the following values:

- FILE_INTERNAL - the file is uploaded/downloaded and stored directly within the Moodle file system.
- FILE_EXTERNAL - the file stays in the external repository and is accessed from there directly.
- FILE_REFERENCE - the file stays in the external repository but may be cached locally. In that case it should be synchronised automatically, as required, with any changes to the external original.

> - FILE_CONTROLLED_LINK - the file remains in the external repository. By "uploading" it, ownership of the file (in the remote system) is changed so that the [system account in the external repository](https://docs.moodle.org//en/OAuth\_2\_services) becomes the new owner of the file. Later, if the file is accessed, the system account is responsible for granting access to users.

Author: done

> - FILE_CONTROLLED_LINK - the file remains in the external repository. By "uploading" it, ownership of the file (in the remote system) is changed so that the [system account in the external repository](https://docs.moodle.org//en/OAuth\_2\_services) becomes the new owner of the file. Later, if the file is accessed, the system account is responsible for granting access to users.

Author: I think I apply all the changes to this part

</Tabs>

The return values influence the choices offered to a user when selecting a file in file picker. Consider the screenshot above, which is from the dialogue that appears directly after choosing (but before uploading) a file in mod_resource. The three options result from FILE_INTERNAL, FILE_REFERENCE, and FILE_CONTROLLED_LINK being present. FILE_REFERENCE corresponds to the "alias/shortcut" option.

The option FILE_EXTERNAL is never reflected in the file picker for mod_resource, so its absence or presence in supported_returntypes() is never reflected here. However, FILE_EXTERNAL is the only return type supported by mod_url: For mod_url, file picker will only(!) list repositories that support FILE_EXTERNAL.

This implies that a plugin that uses a file picker is able to narrow the set of supported return types. For example, assignsubmission_file disallows FILE_EXTERNAL and FILE_REFERENCE.

In the end, which type is
used by Moodle depends on the choices made by the end user (e.g. inserting a link, will result in 'FILE_EXTERNAL'-related functions being used, using a 'shortcut/alias' will result in the 'FILE_REFERENCE'-related functions being used).

#### supported_filetypes()

> Optional method that returns the kind of files a repository can store. Returns '\*' for all file types (default implementation), or an array of types or groups (e.g. array('text/plain', 'image/gif', 'web_image')).

Reviewer: ```suggestion
If your plugin only supports certian file types, then you should implement the optional `supported_filetypes()` method.

This method is used to hide repositories when they don't support certain file types - for example, if a user is inserting a video then any repository which does not support videos will not be shown.

Supported file types can be specified using standard mimetypes (such as `image/gif`) or file groups (such as `web_image`). For a full list of the supported mimetypes and groups, see the [`core_filetypes`](https://github.com/moodle/moodle/blob/v4.0.0/lib/classes/filetypes.php#L47) class.
```

> Optional method that returns the kind of files a repository can store. Returns '\*' for all file types (default implementation), or an array of types or groups (e.g. array('text/plain', 'image/gif', 'web_image')).

Author: I do hate the e.g. this also. I change all of them.

<Tabs>
  <TabItem value="all" label="All files" default>

```php
function supported_filetypes() {
    return '*';
}
```

  </TabItem>
  <TabItem value="mimetypes" label="Filter mimetypes" default>

```php
function supported_filetypes() {
    return ['image/gif', 'image/jpeg', 'image/png'];
}
```

  </TabItem>

>   <TabItem value="groups" label="Filter file groups" default>

Author: Thanks! Copy paste issue. Resolved!

```php
function supported_filetypes() {

>     return ['web_image'];

Reviewer: Recommend adding some comments to the examples.

Reviewer: Also, it's worth noting that you can add return type declaratinos to the child class, that is:
```
function supported_filetypes(): array {
}
```

It may be best if we do so to encourage better type hinting.

Author: Done

}
```

  </TabItem>
</Tabs>

> 
> For a full list of possible types and groups, look in lib/filelib.php, function get_mimetypes_array().

Reviewer: ```suggestion
```

Moved to above the examples. Also this path is no longer accurate. It still exists but doesn't define the available mime types for a few years.

#### Course and User Repository Instances.

> By default, a system wide instance of a repository is created when the repository is enabled.  If you want to allow either or both course specific and user repository instances, you have to specify both **enablecourseinstances** and **enableuserinstances** as options.  There are three ways this can be done:

Reviewer: ```suggestion
A system-wide instance of a repository is created when it is enabled. It is also possible to support both course, and user specific repositories.. This can be achieved by setting the  `enablecourseinstances` and `enableuserinstances` options.  There are three ways that this can be done:
```

> 1. Define **$string\['enablecourseinstances'\]** and **$string\['enableuserinstances'\]** in your plugin's language file.

Reviewer: Make use of backticks.
```suggestion
1. Define `$string['enablecourseinstances']` and `$string['enableuserinstances']` in your plugin's language file.
```

Reviewer: Maybe give examples with comments in the lang file example, i.e.

```
// If your repository supports course or user instances then you should define a language string to explain these.
$string['enablecourseinstances'] = 'Allow administrators to add a [pluginname] repository for a specific course.';
$string['enableuserinstances'] = 'Allow administrators to add a [pluginname] repository for personal use.';
```

> 2. Add them as items in **get_type_option_names()** function's return array.  Note, you must not define the form fields for these options in the **type_config_form()** function.

Reviewer: Bcakticks.

> 3. Several 'core' repositories use the **db/install.php** to create the original repository instance by constructing an instance of the **repository_type** class.  The options can be defined in the array passed as the second parameter to the constructor.

Reviewer: Give an example of one of these in 4.0

Author: I add the repository_filesystem example. This enablecourseinstances and enableuserinstances it is quite strange to me. I documented it as it is right now.

> 

Reviewer: It would be a good idea to create an example, in tabs (one tab per file), in a collapsed section.

> ### Global settings

Reviewer: These aren't settings. Settings are something that a user configures.
This is developer-defined API.

Author: Fine for me

```php
public static function instance_form_validation($mform, $data, $errors) {
    if (empty($data['email_address'])) {
        $errors['email_address'] = get_string('invalidemailsettingname', 'repository_flickrpublic');
    }
}
```

#### Getting / updating settings

Both global and instance settings can be retrieved, from within the plugin, via **$this->get_option('settingname')** and updated via **$this->set_option(array('settingname' => 'value'))**.

:::note

You cannot call **$this** from static methods.  If you need access the non static variables, you may have to store the values in the **_construct()** method into private static variables.

:::

#### plugin_init()

This function must be declared static.

Optional. This function is called when the administrator adds the plugin. So unless the administrator deletes the plugin and re-adds it, it should be called only once.

Parent function does nothing.

### Example of using the settings

As an example, let's create a Flickr plugin for accessing a public flickr account. The plugin will be called "Flickr Public".

Firstly the skeleton:

> ```php title="repository/flickrpublic/lib.php"

Reviewer: The plugin in Moodle has the name `repository_flickr_public`.

Author: Fixed

- **get_type_option_names** returing ['api_key'].
- **type_config_form** adding the api_key text input element into the form.

At this point we have created everything necessary for the administration pages. But let's go further. It would be good if the user can enter any "Flickr public account email address" in the file picker. In fact we want to display in the file picker a Flickr Public repository that the Moodle administrator can never delete. Let's add:

- **plugin_init** using __repository::static_function__ to create a default repository instance.

That's all - the administration part of our Flickr Public plugin is done. For your information, Box.net, Flickr, and Flickr Public all have similar administration APIs.

<details>
  <summary>View the full example code</summary>
  <div>

import FlickPublicLib from '!!raw-loader!./_examples/flickrpublic_lib.php';

> <CodeBlock language="js" title="repository/flickrpublic/lib.php">{FlickPublicLib}</CodeBlock>

Reviewer: It isn't js...

  </div>
</details>

## Repository APIs

### Quick Start

First of all, the File Picker using intensively Ajax you will need a easy way to debug.

- Our first question when you write your plugin specification is 'Does the user need to log-in'? If they do, in your plugin you have to detect user session in constructor() function, and use print_login() if required, see more details below.
- For most of plugins, you need to establish a connection with the remote repository. This connection can be done into the get_listing(), constructor() function, see more details below.
- You want to retrieve the file that the user selected, rewrite get_file() if required, see more details below.

> - Optional question that you should ask yourself is 'Does the user can execute a search', if they do, you will have to rewrite search() method, see more details below.

Reviewer: This section is a very different writing style to the previous. Much more colloqial

Author: Thanks. Fixed.

<details>
  <summary>The full specification of list element</summary>
  <div>

import FullList from '!!raw-loader!./_examples/full_list.php';

<CodeBlock language="php">{FullList}</CodeBlock>

  </div>
</details>

### Dynamically loading

Some repositories contain many files which cannot load in one time, in this case, we need dynamically loading to fetch them step by step, files in subfolder won't be listed until user click the folder in file picker treeview.

As a plug-in developer, if you set dynload flag as **true**, you should return files and folders (set children as a null array) in current path only instead of building the whole file tree.

The use of the **object** tag, instead of returning a **list** of files, allows you to embed an external file chooser within the repository panel. See [[Repository plugins embedding external file chooser]] for details about how to do this.

### User login (optional)

> If the plugin requires login from the user at the time when they use it, then these functions can be used.

Reviewer: add more detail

```suggestion
If our plugin allows a user to log in to a remote service, you can suppor tthis using the `print_login()` and `check_login` function, which are desribed below.
```

Author: done

#### print_login

> Returns an array of the elements required in the login form. If no login form is required, then the default implementation of this will redirect to the files list. If $this->options[is not set, then an HTML-snippet with the login fields (but not the form tags) should be output, instead of returning the form details.

Reviewer: Do we really need this level of detail? It *should* be in the phpdoc.

We shouldn't aim to replace or duplicate it all.

This stanza does not make sense - it looks like part of the sentence is missing (`If $this->options[is not se, then ,...`).

```suggestion
For plugins which need to support login to a remote service, the `print_login()` function can be used to return an array of the form elements needed to support the login.

... (not sure what the rest of the intent of this sentence was to complete it)
```

Author: Ok review this point and clarify the content. I also moved the ajax part to a note.

I noticed that print_login method returns an array or an HTML form depending on the $this->options['ajax'] value. I wouldn't say I like it and probably we should stop doing that but, for now, it should be documented as it is.

<details>
  <summary>View example</summary>
  <div>

```php
public function print_login() { // From repository_pluginname
    global $OUTPUT;

>     if ($this->options['ajax']('ajax'])) {

Reviewer: This line is wrong. `['ajax']('ajax'])`

        $user_field = (object) [

>             'label' => get_string('username', 'repository_pluginname') . ': ',

Reviewer: Mmmn... variable concatenation (I'm guessing unavoidable?)

Author: Most probably it is just copied from an old code. I keep it because it is already there but I will remove it now, we should discourage this kind of hardcoded concatenations to lang specific strings.

            'id'    => 'pluginname_username',
            'type'  => 'text',
            'name'  => 'al_username',
        ];

        $passwd_field = (object) [
            'label' => get_string('password', 'repository_pluginname') . ': ',
            'id'    => 'pluginname_password',
            'type'  => 'password',
            'name'  => 'al_password',
        ];

        $ret = [];
        $ret['login'] = [$user_field, $passwd_field];
        return $ret;
    } else { // Non-AJAX login form - directly output the form elements.
        // Print the login form HTML including the input username and password fields.
        $loginform = new repository_pluginname\output\login();
        echo $OUTPUT->render($loginform);
        // Example of a login form:
        // <label> Username </label>
        // <input type="text" name="al_username" />
        // <label> Password </label>
        // <input type="password" name="al_password" />
        // <input type="submit" value="Enter" />
    }
}
```

  </div>
</details>

This will help to generate a form by file picker which contains user name and password input elements.

> If your login form is static and never changes, you can add __$ret\['allowcaching'\]('login']) = true;__ and filepicker will not send the request to the server every time user opens the login/search form.

Reviewer: backticks.

Author: done

For plugins that do not fully process the login via a popup window, the submitted details can be retrieved, from within the '\__construct' function, via $submitted = optional_param('fieldname', [PARAM_INT/PARAM_TEXT).

<details>
  <summary>View example</summary>
  <div>

> ```php title="Taken from repository_alfresco"

Reviewer: Use the filename instead of this title.
Also, if possible can we use a repo which is in core?

```suggestion
```php title="lib/alfresco/lib.php"
```

  <div>

```php title="Code taken from repository_boxnet"
public function print_login() {
    $ticket = $this->boxclient->getTicket();
    if ($this->options['ajax']) {
        $loginbtn = (object)[
            'type' => 'popup',
            'url' => ' https://www.box.com/api/1.0/auth/' . $ticket->get_oauth_tokens(),
        ];
        $result = [];
        $result['login'] = [$loginbtn];
        return $result;
    } else {
        // Print the login form HTML including the input username, password and ticket fields.
        $loginform = new repository_boxnet\output\login($ticket);
        echo $OUTPUT->render($loginform);
        // Example of a login form:
        // <label> Username </label>
        // <input type="text" name="boxusername" />
        // <label> Password </label>
        // <input type="password" name="boxpassword" />
        // <input type="hidden" name="ticket" value="{$ticket->get_oauth_tokens()}" />
        // <input type="submit" value="Enter" />
    }
}
```

  </div>
</details>

#### check_login

This function will return a boolean value to tell Moodle whether the user has logged in.
By default, this function will return true.

```php

> public function check_login() {

Reviewer: ```suggestion
public function check_login(): bool {
```

Author: done

        // We received a correct header and at least can tell the file size.
        $return = (object) ['filesize' => $curlinfo['download_content_length']];
    }
    return $return;
}
```

  </div>
</details>

#### send_file($storedfile, $lifetime=86400, $filter=0, $forcedownload=false, array $options = null)

Send the requested file back to the user's browser. The 'reference' for the file can be found via $storedfile->get_reference(). If the file is not found / no longer exists, the function 'send_file_not_found()' should be used. Otherwise the file should be output directly, via the most appropriate method:

- Use a 'Location: ' header to redirect to the external URL
- Download the file and cache within the Moodle filesystem (possibly using '$this->import_external_file_contents()'), then call 'send_stored_file'.

:::note

It is up to the repository developer to decide whether to actually download the file or to return a locally cached copy instead.

:::

<details>
  <summary>View example</summary>
  <div>

```php title="Example taken from repository_equella"
public function send_file($stored_file, $lifetime=86400 , $filter=0, $forcedownload=false, array $options = null) {

>     $reference  = unserialize(base64_decode($stored_file->get_reference()));

Reviewer: Yeah, isn't this implemenation a security risk? He URL can be easily fetched and passed to others.

Author: Maybe, I will invent an imaginary method "get_file_details_from_server", it will be more clear

    $url = $this->appendtoken($reference->url);
    if ($url) {
        header('Location: ' . $url);
    } else {
        send_file_not_found();
    }
}
```

  </div>
</details>

An example of caching files within the Moodle filesystem can be found in repository_dropbox.

> ### Misc functions
> 
> A couple of other useful functions to be aware of.
> 

Reviewer: Remove the whole lot.

```suggestion
```

> #### get_name()
> 
> Returns the human-readable name for this instance of the plugin (the default implementation should usually be fine and this function can be useful when doing any output to the user).
> 

Reviewer: If the default is fine for 99% of people, remove this.
```suggestion
```

#### cron()

> For any background tasks that need to be scheduled (rarely needed). The minimum time between calls is specified in the version.php file (but the maximum time depends on the server settings for the Moodle install).

Author: done
